### ISSUE TITLE
PHP7 compatible release

### ISSUE THREAD

**Issue by phadej** [igbinary/igbinary7#1]
https://github.com/igbinary/igbinary/issues/40

**Comment by phadej** [comment]
Remember:
![opensource](https://cloud.githubusercontent.com/assets/51087/11867171/0b149a34-a4b8-11e5-9fb1-a6ea3a7270af.jpg)

**Comment by phadej** [comment]
There seems to be failing tests related to cyclic/reference stuff. I heard that there're changes in PHP, but I still don't understand them properly.

**Comment by krakjoe** [comment]
https://github.com/igbinary/igbinary7/blob/master/igbinary.c#L1533

Here, the `zval*` passed to `igbinary_serialize_array_ref` is always going to be `IS_REFERENCE` type.

I _think_ it should pass `&Z_REFVAL_P(z)` like `if (igbinary_serialize_array_ref(igsd, &Z_REFVAL_P(z), false TSRMLS_CC) == 0)`

I can see `ALLOC_INIT_ZVAL`, bit of inconsistency between current master and this branch in unserialize and serialize routines for zvals ...

It generally wants just a review ... I'll have the time to do it myself some time next week ... probably ...

**Comment by tricky** [comment]
@krakjoe  Oh! absolutely in need of a review, _especially_ regarding refs. most likely needs some rewriting. Without refs we're quite safe though.

But anyway eyes on the prize, and a fully working release is it.

**Comment by phadej** [comment]
Failing tests atm:

```
Check for reference serialisation [tests/igbinary_009.phpt]
Object-Reference test [tests/igbinary_014.phpt]
Cyclic array test [tests/igbinary_026.phpt]
Unserialize backwards compatible with v1. [tests/igbinary_unserialize_v1_compatible.phpt]
```

@tricky Do you have some failing fixtures to add?

_EDIT_ those tests aren't probably the best ones, as we should have more or less only

``` php
assert_equal(igbinary_unserialise(igbinary_serialise($data)), $data);
```

for some definition of `assert_equal`.

What is the right definition is an actual mystery in presence of references.

**Comment by phadej** [comment]
I have to revisit @tricky's commit in https://github.com/igbinary/igbinary/tree/php7-dev-playground2

It seems that `igbinary7` doens't compile for me locally (I don't know how it compiles on travis - suspicious - picks some other php version headers?); will check that during weekend.

**Comment by Sean-Der** [comment]
Hi @phadej 

I have ported/released everything to PECL that I need for my stuff to be _functional_. So going to work on igbinary now until all the tests are fixed. Two questions
- Why did you decide to use the fork you did? It has less passing tests than others iirc, is the code cleaner (haven't looked yet)
- You and @tricky aren't working on this at the moment? I don't want to duplicate work again.

As soon as this is good to go I will start hacking on this again! I will also bump the protocol if needed. thanks!

**Comment by Spea** [comment]
I'm curious if there is any progress here, would be nice if you can give an update about this :)

**Comment by dfreudenberger** [comment]
i'll just leave a comment to get updates on this as well

**Comment by RdeWilde** [comment]
Good efforts!

**Comment by razvanphp** [comment]
Currently this is the reason we're not switching to PHP7 yet.

I understand the idea about open source, but there seem to be a lot of people willing to contribute that are somehow blocked by the ones with commit rights.

If maintainers have no time for this, you should consider giving those rights to other interested and talented people so the project can go forward.

PHP7 was lunched 6 months ago to GA, and was in beta even more time before, people need igbinary!

Thank you!

**Comment by jeffreyddavis** [comment]
I would like to move to PHP 7 as well, but igbinary is a critical component of our infrastructure. Microsoft will likely be releasing the SQL server driver for PHP7 in June. So that might be a good target date, since many people will want to move to php7 then.

**Comment by dkeepersun** [comment]
+100

**Comment by GuillaumeRossolini** [comment]
Hi,

I know there are a few issues remaining (for example https://github.com/igbinary/igbinary/issues/58 mistakenly reported in the other repo).
Would it help resolving that issue if I provided a simple reproduce case? Although @brstgt seems to have already done that.

Thanks,

**Comment by RahimovIR** [comment]
@Kirkis you can try to fix https://github.com/igbinary/igbinary7/pull/8

**Comment by GuillaumeRossolini** [comment]
@RahimovIR Sorry for the misunderstanding _[edit: my previous comment made me seem competent at extension development, which I am not; I censored it]_.
Nice to know there is a reproduce case already, though.

**Comment by RdeWilde** [comment]
Awesome!

**Comment by TysonAndre** [comment]
https://github.com/igbinary/igbinary7/issues/16 - there are some edge cases in unserialize if an object implements __wakeup and in __wakeup, if properties that are references, arrays, or objects are reassigned or modified. Or if __wakeup adds dynamic properties. Those should be fixed before an alpha.

**Comment by TysonAndre** [comment]
After https://github.com/igbinary/igbinary7/pull/17 is merged, I think this should be ready for an alpha /beta release

@phadej @tricky - I'm pretty sure that #17 is ready, but wanted a review for that.

Aside: Edge cases in __wakeup also existed outside of igbinary: https://bugs.php.net/bug.php?id=72610

**Comment by TysonAndre** [comment]
Alpha candidate will probably be https://github.com/igbinary/igbinary7/tree/alpha-candidate-2016-07-18 (includes #17 and #18), plus nits on replacing ulong with zend_ulong

If there are no bugs found, I plan to merge this by the end of the week.

Edit: merged to master. The remaining tasks I can think of for igbinary7 are implementing performance improvements (`hash_si`\* data structure/algorithm, re-using `zend_string`s in both values and (in php7) keys when unserializing)

**Comment by TysonAndre** [comment]
Created https://github.com/igbinary/igbinary7/issues/19

**Comment by razvanphp** [comment]
@TysonAndre thank you for your great work, can we assume the master of igbinary repo is the latest php7 release we should use? is it production ready?

**Comment by TysonAndre** [comment]
use https://github.com/igbinary/igbinary instead of the igbinary7 repo.
If I saw any bugs related to serialization/unserialization, fixes (and tests of fixes) would be merged to igbinary/igbinary promptly.

The serialization/unserialization is well tested, and I've been running it in production with memcache for a while now (with a mix of php and php servers).

I'm personally using https://github.com/TysonAndre/igbinary master branch (this contains fixes for integration with other extensions, as well as performance improvements for serialization. Those changes haven't been reviewed by other maintainers yet. )

https://github.com/igbinary/igbinary/pull/65 (APCU integration) and https://github.com/igbinary/igbinary/pull/66 (Session serializer integration) will be merged on monday, if there are no review comments from other maintainers.
- Memcached+igbinary is working reliably for me. PHP7 serialized objects can be read by php5, and vice versa.
- APCU(+igbinary) has PRs to fix bugs (some related to igbinary, some not). They aren't merged yet. See https://github.com/krakjoe/apcu/pulls (And TysonAndre/apcu for what I use).
  EDIT: Those PRs are merged, custom serializer bugs in https://github.com/krakjoe/apcu/ should be fixed.
- Session handler(+igbinary serializer) - Haven't personally tested it, but the serialization/unserialization is the same as what memcache uses, so it should be as reliable as memcached+igbinary.

**Comment by TysonAndre** [comment]
https://github.com/igbinary/igbinary/issues/86 is a duplicate of this, and more visible and recently active.

https://github.com/igbinary/igbinary/releases/tag/2.0.0 is what is proposed as a stable release.
